platform/pulp/platform/models/storage.py

```diff
@@ -0,0 +1,101 @@
+import os
+import errno
+import shutil
+
+from datetime import datetime
+
+from django.conf import settings
+from django.core.files import File
+from django.core.files.storage import Storage
+from django.utils.deconstruct import deconstructible
+
+
+@deconstructible
+class StoragePath(object):
+
+    @staticmethod
+    def base_path(model):
+        digest = model.content.natural_key_digest()
+        return os.path.join(
+            settings.MEDIA_ROOT,
+            'units',
+            model.content.type,
+            digest[0:2],
+            digest[2:])
+
+    def __call__(self, model, name):
+        return os.path.join(StoragePath.base_path(model), model.published_path)
+
+
+class FileSystem(Storage):
+
+    @staticmethod
+    def mkdir(path):
+        try:
+            os.makedirs(path)
+        except OSError as e:
+            if e.errno != errno.EEXIST:
+                raise
+
+    @staticmethod
+    def link(path, target):
+        FileSystem.mkdir(os.path.dirname(path))
+        os.symlink(path, target)
+
+    @staticmethod
+    def unlink(path):
+        try:
+            os.unlink(path)
+        except OSError as e:
+            if e.errno != errno.ENOENT:
+                raise
+
+    @staticmethod
+    def _open(path, mode='rb'):
+        return File(open(path, mode))
+
+    @staticmethod
+    def _save(path, content):
+        FileSystem.mkdir(os.path.dirname(path))
+        shutil.copy(content.name, path)
+        return path
+
+    def get_available_name(self, name, max_length=None):
+        if max_length:
+            if len(name) > max_length:
+                raise ValueError('max_length exceeded')
+        return name
+
+    def delete(self, path):
+        FileSystem.unlink(path)
+
+    def exists(self, path):
+        return os.path.exists(path)
+
+    def listdir(self, path):
+        files = []
+        directories = []
+        for entry in os.listdir(path):
+            if os.path.isdir(os.path.join(path, entry)):
+                directories.append(entry)
+            else:
+                files.append(entry)
+        return (directories, files)
+
+    def path(self, path):
+        return path
+
+    def size(self, path):
+        return os.path.getsize(path)
+
+    def url(self, path):
+        'file://{}'.format(path)
```

Comment:
According to the `Storage.url()` docstring - "Returns an absolute URL where the file's contents can be accessed directly by a Web browser."  This implementation probably needs a bit more work and likely needs to incorporate `settings.MEDIA_URL`.  I also don't know how this get's used by django.